kernel/kernel-api/src/test/scala/io/delta/kernel/internal/util/ColumnMappingSuite.scala

```diff
@@ -515,6 +519,93 @@ class ColumnMappingSuite extends AnyFunSuite with ColumnMappingSuiteBase {
+      (
+        new StructType().add(
+          "p",
+          baseSchema,
+          /* nullable= */ false),
+        2,
+        (md: Metadata) =>
+          md.getSchema.get("p").getDataType.asInstanceOf[StructType].get("l"))).foreach {
+      case (schemaToTest, base, getter) =>
+
+        test(s"Deeply nested values don't assign more field IDs then necessary $isNewTable $base") {
+
+          var inputMetadata = testMetadata(schemaToTest).withColumnMappingEnabled("id")
+          inputMetadata = inputMetadata.withIcebergCompatV2Enabled
+          inputMetadata = inputMetadata.withIcebergWriterCompatV1Enabled
+          val metadata = updateColumnMappingMetadataIfNeeded(inputMetadata, isNewTable)
+            .orElseGet(() => fail("Metadata should not be empty"))
+
+          assertThat(metadata.getConfiguration).containsEntry(
+            ColumnMapping.COLUMN_MAPPING_MAX_COLUMN_ID_KEY,
+            (base + 8).toString)
+          val prefix = s"col-$base"
+          // Values are offset by base.  All Ids are assigned in depth first order to
+          // StructField's first and then intermediate nested fields are added after.
+          val nestedColumnMappingValues = FieldMetadata.builder()
+            .putLong(s"$prefix.element", base + 2)
+            .putLong(s"$prefix.element.element", base + 3)
+            .putLong(s"$prefix.element.element.key", base + 4)
+            .putLong(s"$prefix.element.element.key.element", base + 5)
+            .putLong(s"$prefix.element.element.value", base + 6)
+            .putLong(s"$prefix.element.element.value.key", base + 7)
+            .putLong(s"$prefix.element.element.value.value", base + 8).build()
+          val expectedMetadata = FieldMetadata.builder().putFieldMetadata(
+            COLUMN_MAPPING_NESTED_IDS_KEY,
+            nestedColumnMappingValues)
+            .putString("delta.columnMapping.physicalName", prefix)
+            .putLong("delta.columnMapping.id", base)
+            .putBoolean("k2", true).build()
+          val firstColumnMetadata = getter(metadata).getMetadata
+          assertThat(firstColumnMetadata.getMetadata(
+            COLUMN_MAPPING_NESTED_IDS_KEY).getEntries).containsExactlyInAnyOrderEntriesOf(
+            nestedColumnMappingValues.getEntries)
+          assertThat(firstColumnMetadata.getEntries).containsExactlyInAnyOrderEntriesOf(
+            expectedMetadata.getEntries)
+
+          // TODO: It would be nice to have visitor pattern on schema so we can assert
+          // all metadata for nested fields
+          // are empty but this at least provides a sanity check.
+          assert(getter(metadata).getDataType.asInstanceOf[
+            ArrayType].getElementField.getMetadata == FieldMetadata.empty())
+
+          // Requesting the same operation on the same schema shouldn't change anything
+          // as the schema already has the necessary column mapping info
+          assertNoOpOnUpdateColumnMappingMetadataRequest(
```

Comment:
If I follow this assertion implicitly validates that if pre-assigned IDs were passed to the schema that's being set, that we're not doing any column mapping update or max field ID update right?  Basically asserting we are hitting https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java#L380 as expected

Reply:
+1 also wondered the same thing - maybe add a comment "this also validates that the maxFieldId is not updated" just for future clarity?